sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -936,11 +944,26 @@ public Read<K, V> withTopics(List<String> topics) {
      */
     public Read<K, V> withTopicPartitions(List<TopicPartition> topicPartitions) {
       checkState(
-          getTopics() == null || getTopics().isEmpty(),
-          "Only topics or topicPartitions can be set, not both");
+          (getTopics() == null || getTopics().isEmpty()) && getTopicPattern() == null,
+          "Only one of topics, topicPartitions or topicPattern can be set");
       return toBuilder().setTopicPartitions(ImmutableList.copyOf(topicPartitions)).build();
     }
 
+    /**
+     * Sets a {@link java.util.regex.Pattern} of topics to read from. All the partitions from each
+     * of the matching topics are read.
+     *
+     * <p>See {@link KafkaUnboundedSource#split(int, PipelineOptions)} for description of how the
+     * partitions are distributed among the splits.
+     */
+    public Read<K, V> withTopicPattern(Pattern topicPattern) {
```

Comment:
Can change this API to use a portable type (for example, a string regex) so that this can be supported via cross-language wrappers ?

Reply:
`Read.External.Configuration` has fields for `keyDeserializer` and `valueDeserializer` which are resolved from `String` to `Class` during external transform construction, does it make sense to provide a mapping there instead?

Reply:
Ideally we don't want to be using external read configuration as a pattern. It makes it so the configurations for an IO diverge for python v. java users

Reply:
Got it. Would an additional method overload with `String topicPattern` help at all?
Otherwise I'll change the method signature.

Reply:
I think we can have both

Reply:
Thanks. If there is no significant advantage for specifying a Pattern object I would just support specifying a String regex to keep the API simple.

Reply:
Done.